content/influxdb/cloud-serverless/process-data/downsample_quix.md

```diff
@@ -0,0 +1,333 @@
+```
+
+## Prepare InfluxDB buckets
+
+The downsampling process involves two InfluxDB buckets.
+Each bucket has a [retention period](/influxdb/cloud-serverless/reference/glossary/#retention-period)
+that specifies how long data persists in the database before it expires and is deleted.
+By using two buckets, you can store unmodified, high-resolution data in a bucket
+with a shorter retention period and then downsampled, low-resolution data in a
+bucket with a longer retention period.
+
+Ensure you have a bucket for each of the following:
+
+- One to query unmodified data from
+- The other to write downsampled data to
+
+For information about creating buckets, see
+[Create a bucket](/influxdb/cloud-serverless/admin/buckets/create-bucket/).
+
+<!--Starting with downsampling becuase data I/O is fairly mundane and not the main focus of the tutorial -->
+## Create the downsampling logic
+
+This process reads the raw data from the input Kafka topic (which stores data streamed from InfluxDB), downsamples it, the sends it to an output topic (which is used to write back to InfluxDB)
+
+First, use the Quix Streams library's `Application` class to initialize a connection to Apache Kafka.
+
+```py
+from quixstreams import Application
+from quixstreams.models.serializers.quix import JSONDeserializer, JSONSerializer
+
+app = Application.Quix(consumer_group="downsampling-process", auto_offset_reset="earliest")
+input_topic = app.topic('raw-data', value_deserializer=JSONDeserializer())
+output_topic = app.topic('downsampled-data', value_serializer=JSONSerializer())
+
+## ... remaining code trunctated for brevity ...
+```
+
+Then, configure the Quix Streams built-in windowing function to create a tumbling window to continously downsample the data into 1-minute buckets. 
+
+```py
+...
+target_field = 'temperature' # The field that you want to downsample.
+
+def custom_ts_extractor(value):
+    # ...truncated for brevity - custom code that defines the 'time_recorded' field as the timestamp to use for windowing...
+
+topic = app.topic(input_topic, timestamp_extractor=custom_ts_extractor)
+
+sdf = (
+    sdf.apply(lambda value: value[target_field])  # Extract temperature values
+    .tumbling_window(timedelta(minutes=1))   # 1-minute tumbling windows
+    .mean()                                  # Calculate average temperature
+    .final()                                 # Emit results at window completion
+)
+
+sdf = sdf.apply(
+    lambda value: {
+        "time": value["end"],                  # End of the window
+        "temperature_avg": value["value"],     # Average temperature
+    }
+)
+
+sdf.to_topic(output_topic) # Output results to the 'downsampled-data' topic
+...
+```
+The results are streamed to the Kafka topic `downsampled-data`.
+
+Note: "sdf" stands for "Streaming Dataframe".
+
+* You can find the full code for this process in the [Quix GitHub repository](https://github.com/quixio/template-influxdbv3-downsampling/blob/dev/Downsampler/main.py).
```

Comment:
~~~suggestion
1.  Use the Quix Streams library's `Application` class to initialize a connection to Apache Kafka.

    ```py
    from quixstreams import Application
    from quixstreams.models.serializers.quix import JSONDeserializer, JSONSerializer

    app = Application.Quix(consumer_group="downsampling-process", auto_offset_reset="earliest")
    input_topic = app.topic('raw-data', value_deserializer=JSONDeserializer())
    output_topic = app.topic('downsampled-data', value_serializer=JSONSerializer())

    # ...
    ```

2.  Configure the Quix Streams built-in windowing function to create a tumbling
    window that continously downsamples the data into 1-minute buckets. 

    ```py
    # ...
    target_field = 'temperature' # The field that you want to downsample.

    def custom_ts_extractor(value):
        # ...
        # truncated for brevity - custom code that defines the 'time_recorded'
        # field as the timestamp to use for windowing...

    topic = app.topic(input_topic, timestamp_extractor=custom_ts_extractor)

    sdf = (
        sdf.apply(lambda value: value[target_field])  # Extract temperature values
        .tumbling_window(timedelta(minutes=1))   # 1-minute tumbling windows
        .mean()                                  # Calculate average temperature
        .final()                                 # Emit results at window completion
    )

    sdf = sdf.apply(
        lambda value: {
            "time": value["end"],                  # End of the window
            "temperature_avg": value["value"],     # Average temperature
        }
    )

    sdf.to_topic(output_topic) # Output results to the 'downsampled-data' topic
    # ...
    ```

The results are streamed to the Kafka topic, `downsampled-data`.

{{% note %}}
Note: "sdf" stands for "Streaming Dataframe".
{{% /note %}}

You can find the full code for this process in the
[Quix GitHub repository](https://github.com/quixio/template-influxdbv3-downsampling/blob/dev/Downsampler/main.py).
~~~